### PR TITLE
Split subsample streamline

### PR DESCRIPTION
**Pull request by arnaudbore**
scil_subsample has been split into 3 new scripts:

- scil_filter_by_length_streamlines.py
- scil_get_subset_streamlines.py
- scil_resample_streamlines.py

### CODE REVIEW

File: scilpy/tracking/tools.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                                 data_per_point,
+                                 data_per_streamline,
+                                 min_length=0., max_length=np.inf):
+    """
+    Parameters
+    ----------
+    streamlines: list
+        List of list of 3D points.
+
+    data_per_point: dict
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: Please add a small description for data per point and streamline

File: scilpy/tracking/tools.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
+from __future__ import division
+
+import dipy.tracking.metrics
+from dipy.tracking.streamline import set_number_of_points
+import dipy.tracking.utils
+import numpy as np
+
+
+def filter_streamlines_by_length(streamlines,
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: Add a small documentation of what the function do.

File: scilpy/tracking/tools.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    per_point = data_per_point[filterStream]
+    per_streamline = data_per_streamline[filterStream]
+
+    data = {'per_point': per_point,
+            'per_streamline': per_streamline}
+
+    return list(np.asarray(streamlines)[filterStream]), data
+
+
+def get_subset_streamlines(streamlines,
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: Add a small documentation of what the function do.

File: scilpy/tracking/tools.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                           data_per_streamline,
+                           max_streamlines, rng=None):
+    """
+    Parameters
+    ----------
+    streamlines: list
+        List of list of 3D points.
+    max_streamlines: int
+        Maximum number of streamlines to output.
+    rng: RandomState object
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: rng_seed

File: scilpy/tracking/tools.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        Random number generator to use for shuffling the data.
+        By default, a constant seed is used.
+
+    Return
+    ------
+    average: list
+        List of subsampled streamlines.
+    """
+
+    if rng is None:
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: else create rng with rng_seeds

File: scilpy/tracking/tools.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    """
+
+    num_streamlines = len(streamlines)
+
+    lengths = np.zeros(num_streamlines)
+
+    for i in np.arange(num_streamlines):
+        lengths[i] = dipy.tracking.metrics.length(streamlines[i])
+
+    filterStream = np.logical_and(lengths >= min_length, lengths <= max_length)
```
Comment (frheault) [inline scilpy/tracking/tools.py]: filter_stream

File: scripts/scil_filter_by_length_streamlines.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    parser = _build_args_parser()
+    args = parser.parse_args()
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    assert_inputs_exist(parser, [args.input])
+    assert_outputs_exists(parser, args, args.output)
+
+    tractogramFile = load(args.input)
```
Comment (frheault) [inline scripts/scil_filter_by_length_streamlines.py]: tractogram_file

File: scilpy/tracking/tools.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        List of list of 3D points.
+    max_streamlines: int
+        Maximum number of streamlines to output.
+    rng: RandomState object
+        Random number generator to use for shuffling the data.
+        By default, a constant seed is used.
+
+    Return
+    ------
+    average: list
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: Add data_per_point and data_per_streamline in return

File: scripts/scil_resample_streamlines.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    parser = _build_args_parser()
+    args = parser.parse_args()
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    assert_inputs_exist(parser, [args.input])
+    assert_outputs_exists(parser, args, args.output)
+
+    tractogramFile = load(args.input)
```
Comment (frheault) [inline scripts/scil_resample_streamlines.py]: tractogram_file

File: scilpy/tracking/tools.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    ind = np.arange(len(streamlines))
+    rng.shuffle(ind)
+
+    per_point = data_per_point[ind[:max_streamlines]]
+    per_streamline = data_per_streamline[ind[:max_streamlines]]
+
+    data = {'per_point': per_point,
+            'per_streamline': per_streamline}
+
+    return list(np.asarray(streamlines)[ind[:max_streamlines]]), data
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: same than before split data in two

File: scripts/scil_get_subset_streamlines.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    parser = _build_args_parser()
+    args = parser.parse_args()
+
+    if args.verbose:
+        logging.basicConfig(level=logging.INFO)
+
+    assert_inputs_exist(parser, [args.input])
+    assert_outputs_exists(parser, args, args.output)
+
+    tractogramFile = load(args.input)
```
Comment (frheault) [inline scripts/scil_get_subset_streamlines.py]: tractogram_file

File: scilpy/tracking/tools.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    arc_length: bool
+        Whether to downsample using arc length parametrization.
+
+    Return
+    ------
+    average: list
+        List of subsampled streamlines.
+    """
+    results = []
+    for i in range(len(streamlines)):
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: for streamline in streamlines

File: scilpy/tracking/tools.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    streamlines: list
+        List of list of 3D points.
+    num_points: int
+        Number of points per streamline in the output.
+    arc_length: bool
+        Whether to downsample using arc length parametrization.
+
+    Return
+    ------
+    average: list
```
Comment (GuillaumeTh) [inline scilpy/tracking/tools.py]: Please modify doc

File: scripts/scil_filter_by_length_streamlines.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+from scilpy.tracking.tools import filter_streamlines_by_length
+from scilpy.io.utils import (assert_inputs_exist, assert_outputs_exists,
+                             add_overwrite_arg)
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Subsample a set of streamlines.\n'
+                    'WARNING: data_per_point is carried')
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: Remove warning message

File: scripts/scil_get_subset_streamlines.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+from scilpy.tracking.tools import get_subset_streamlines
+from scilpy.io.utils import (assert_inputs_exist, assert_outputs_exists,
+                             add_overwrite_arg)
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Subsample a set of streamlines.\n'
+                    'WARNING: data_per_point is not carried')
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: Remove warning message

File: scripts/scil_filter_by_length_streamlines.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: scil_filter_streamlines_by_length.py

File: scripts/scil_filter_by_length_streamlines.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+from scilpy.tracking.tools import filter_streamlines_by_length
+from scilpy.io.utils import (assert_inputs_exist, assert_outputs_exists,
+                             add_overwrite_arg)
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Subsample a set of streamlines.\n'
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: Please change the description

File: scripts/scil_filter_by_length_streamlines.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+                    'WARNING: data_per_point is carried')
+    p.add_argument(
+        'input', action='store',  metavar='input',
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: remove action, metavar and type

File: scripts/scil_filter_by_length_streamlines.py [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        type=str,  help='Streamlines input file name.')
+    p.add_argument(
+        'output', action='store',  metavar='output',
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: same

File: scripts/scil_filter_by_length_streamlines.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        type=str,  help='Streamlines output file name.')
+    p.add_argument(
+        '--minL', default=0., type=float,
+        help='Minimum length of streamlines. [%(default)s]')
+    p.add_argument(
+        '--maxL', default=0., type=float,
+        help='Maximum length of streamlines. [%(default)s]')
+    p.add_argument('-v', action='store_true', dest='verbose',
```
Comment (GuillaumeTh) [inline scripts/scil_filter_by_length_streamlines.py]: Use the function

File: scripts/scil_get_subset_streamlines.py [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+from scilpy.tracking.tools import get_subset_streamlines
+from scilpy.io.utils import (assert_inputs_exist, assert_outputs_exists,
+                             add_overwrite_arg)
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Subsample a set of streamlines.\n'
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: Change description

File: scripts/scil_get_subset_streamlines.py [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+                    'WARNING: data_per_point is not carried')
+    p.add_argument(
+        'input', action='store',  metavar='input',
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: same

File: scripts/scil_get_subset_streamlines.py [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+        type=str,  help='Streamlines input file name.')
+    p.add_argument(
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: same

File: scripts/scil_get_subset_streamlines.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        'output', action='store',  metavar='output',
+        type=str,  help='Streamlines output file name.')
+    p.add_argument(
+        '-n', dest='n', action='store', metavar=' ', default=0,
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: remove dest, store, metavar

File: scripts/scil_get_subset_streamlines.py [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+        type=int, help='Maximum number of streamlines to output. [all]')
+    p.add_argument('-v', action='store_true', dest='verbose',
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: Use function

File: scripts/scil_resample_streamlines.py [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+                             add_overwrite_arg)
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Subsample a set of streamlines.\n'
+                    'WARNING: data_per_point is not carried')
+    p.add_argument(
+        'input', action='store',  metavar='input',
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: same

File: scripts/scil_resample_streamlines.py [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+        type=str,  help='Streamlines input file name.')
+    p.add_argument(
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: same

File: scripts/scil_resample_streamlines.py [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+        'output', action='store',  metavar='output',
+        type=str,  help='Streamlines output file name.')
+    p.add_argument(
+        '--npts', dest='npts', action='store', metavar=' ',
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: remove dest, action, metavar

File: scripts/scil_resample_streamlines.py [hunk 29]
```diff
@@ -0,0 +1,69 @@
...
+        default=0, type=int,
+        help='Number of points per streamline in the output. [%(default)s]')
+    p.add_argument(
+        '--arclength', dest='arclength', action='store_true', default=False,
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: remove dest.

File: scripts/scil_resample_streamlines.py [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+        help='Whether to downsample using arc length parametrization. ' +
+             '[%(default)s]')
+
+    p.add_argument('-v', action='store_true', dest='verbose',
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: Use function

File: scripts/scil_get_subset_streamlines.py [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+                   help='Produce verbose output. [%(default)s]')
+
+    add_overwrite_arg(p)
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: Add option --seed to set the rng seed to use

File: scripts/scil_resample_streamlines.py [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Resample a set of streamlines.\n'
+                    'WARNING: data_per_point is not carried')
+    p.add_argument('in_tractogram',
+                   help='Streamlines input file name.')
+    p.add_argument('nb_pts_per_streamline', type=int,
+                   help='Number of points per streamline in the output. [%(default)s]')
```
Comment (GuillaumeTh) [inline scripts/scil_resample_streamlines.py]: Please remove default in help

File: scripts/scil_get_subset_streamlines.py [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def _build_args_parser():
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description='Get a subset of streamlines.')
+    p.add_argument('in_tractogram',
+                   help='Streamlines input file name.')
+    p.add_argument('max_num_streamlines', type=int,
+                   help='Maximum number of streamlines to output. [all]')
```
Comment (GuillaumeTh) [inline scripts/scil_get_subset_streamlines.py]: please remove [all]

**Pull request merged**